File: src/main/kotlin/_2_null_safety/exercise/null_allowed/NullableTypes.java
package _2_null_safety.exercise.null_allowed;

public class NullableTypes {
    public void sendMessageToClient(
            ClientJ client,
            String message,
            MailerJ mailer
    ) {
        if (client == null || message == null) return;

        PersonalInfoJ personalInfo = client.getPersonalInfo();
        if (personalInfo == null) return;

        String email = personalInfo.getEmail();
        if (email == null) return;

        mailer.sendMessage(email, message);
    }


    public static void main(String[] args) {
        NullableTypes nullableTypes = new NullableTypes();

        ClientJ mail = new ClientJ(new PersonalInfoJ("Mail"));
        nullableTypes.sendMessageToClient(mail, "Java hello kata", (email, message) -> System.out.println(message));
    }
}

class ClientJ {
    private final PersonalInfoJ personalInfo;

    public ClientJ(PersonalInfoJ personalInfo) {
        this.personalInfo = personalInfo;
    }

    public PersonalInfoJ getPersonalInfo() {
        return personalInfo;
    }
};

class PersonalInfoJ {
    private final String email;

    public PersonalInfoJ(String email) {
        this.email = email;
    }

    public String getEmail() {
        return email;
    }
};

interface MailerJ {
    void sendMessage(String email, String message);
}

File: src/main/kotlin/_2_null_safety/exercise/null_allowed/NullableTypes.kt
package _1_basic_syntax.exercise.nullable_types

/**
 * Look at sendMessageToClient in NullableTypes.java,
 * Using kotlin built in null safety it can be rewritten to single if statement
 */
fun sendMessageToClient(
    client: Client?, message: String?, mailer: Mailer
) {
    TODO()
}

class Client(val personalInfo: PersonalInfo?)
class PersonalInfo(val email: String?)
interface Mailer {
    fun sendMessage(email: String, message: String)
}

fun main() {
    val client = Client(PersonalInfo("email"))
    val anonymousMailer = object : Mailer {
        override fun sendMessage(email: String, message: String) = println("Sending message to $email")
    }
    sendMessageToClient(client, "hey kata", anonymousMailer)
}
